csrc/quantization/compressed_tensors/int8_quant_kernels.cu

```diff
@@ -103,134 +105,165 @@ static inline __device__ int8_t int32_to_int8(int32_t x) {
+template <typename scalar_t, typename scale_t, typename azp_t>
 __global__ void dynamic_scaled_int8_azp_quant_kernel(
-    scalar_t const* __restrict__ input, int8_t* __restrict__ out,
-    scale_type* scale, azp_type* azp, const int hidden_size) {
-  int64_t const token_idx = blockIdx.x;
-
-  // Must be performed using 64-bit math to avoid integer overflow.
-  out += token_idx * hidden_size;
-  input += token_idx * hidden_size;
-
-  // Scan for the min and max value for this token
-  float max_val = std::numeric_limits<float>::min();
-  float min_val = std::numeric_limits<float>::max();
-  for (int i = threadIdx.x; i < hidden_size; i += blockDim.x) {
-    auto val = static_cast<float>(input[i]);
-    max_val = std::max(max_val, val);
-    min_val = std::min(min_val, val);
+    const scalar_t* __restrict__ input, int8_t* __restrict__ output,
+    scale_t* scale_out, azp_t* azp_out, const int hidden_size) {
+  int tid = threadIdx.x;
+  int stride = blockDim.x;
+
+  const scalar_t* row_in = input + blockIdx.x * hidden_size;
+  int8_t* row_out = output + blockIdx.x * hidden_size;
+
+  // 1. calculate min & max
+  float tmin = FLT_MAX;
+  float tmax = -FLT_MAX;
+  for (int i = tid; i < hidden_size; i += stride) {
+    float v = float(row_in[i]);
+    tmin = fminf(tmin, v);
+    tmax = fmaxf(tmax, v);
   }
 
-  // Reduce the max and min values across the block
-  using BlockReduce = cub::BlockReduce<float, 1024>;
-  __shared__ typename BlockReduce::TempStorage reduceStorage;
-  max_val = BlockReduce(reduceStorage).Reduce(max_val, cub::Max{}, blockDim.x);
-  __syncthreads();  // Make sure min doesn't mess with max shared memory
-  min_val = BlockReduce(reduceStorage).Reduce(min_val, cub::Min{}, blockDim.x);
-
-  __shared__ scale_type scale_sh;
-  __shared__ azp_type azp_sh;
-
-  // Compute the scale and zero point and store them, only on the first thread
-  if (threadIdx.x == 0) {
-    float const scale_val = (max_val - min_val) / 255.0f;
-    // Use rounding to even (same as torch.round)
-    auto const azp_float = std::nearbyint(-128.0f - min_val / scale_val);
-    auto const azp_val = static_cast<azp_type>(azp_float);
-
-    // Store the scale and azp into shared and global
-    scale[token_idx] = scale_sh = scale_val;
-    azp[token_idx] = azp_sh = azp_val;
+  // This is used for BlockReduce
+  struct MinMax {
+    float mn, mx;
+  };
+  struct ReduceMinMax {
+    __device__ MinMax operator()(const MinMax& a, const MinMax& b) const {
+      return {fminf(a.mn, b.mn), fmaxf(a.mx, b.mx)};
+    }
+  };
+
+  using BlockReduce = cub::BlockReduce<MinMax, 256>;
+  __shared__ typename BlockReduce::TempStorage tmp;
+
+  MinMax thread_mm{tmin, tmax};
+  MinMax mm = BlockReduce(tmp).Reduce(thread_mm, ReduceMinMax(), blockDim.x);
+  __shared__ float scale_sh;
+  __shared__ azp_t azp_sh;
+  if (tid == 0) {
+    float s = (mm.mx - mm.mn) / 255.f;
+    float zp = nearbyintf(-128.f - mm.mn / s);  // round-to-even
```

Comment:
I think this is quite hard to read, why not keep the `max` and `min` member names?